### PR TITLE
Add more PWscf parsing rules and unit tests

### PR DESCRIPTION
**Pull request by hegdevinayi**
Adding new rules for parsing the following properties from PWscf output:
- name of the input file
- cell vectors
- number of atoms in the unit-cell
- number atom types
- number of electrons
- mixing beta
- scf convergence threshold
- ionic convergence thresholds (energy/forces/stress)
- k-points and weights
- smearing type and width
- Hubbard energy contribution
- atomic positions
- total and absolute magnetization
- site-projected charges and magnetic moments
- total CPU time
- atomic positions (initial and updated)

Minor changes to the parsing rules for:
- PWscf header (version/date/time information)
- total energy (parse only the self-consistent energies)

Some minor refactoring of function names for clarity (e.g. `_parse_bfgs`-> `_parse_n_bgfs_steps`).

### CODE REVIEW

File: dftparse/pwscf/stdout_parser.py [hunk 1]
```diff
@@ -1,173 +1,481 @@
...
+    (lambda x: 'Magnetic moment per site' in x, _parse_site_proj_quantities),
+    (lambda x: 'warning' in x.lower(), _parse_warning),
 ]
 
+
 class PwscfStdOutputParser(BlockParser):
 
     def __init__(self, rules=base_rules):
-        BlockParser.__init__(self)
+        super(PwscfStdOutputParser, self).__init__()
```
Comment (maxhutch) [inline dftparse/pwscf/stdout_parser.py]: Why this change?  They look equivalent to me, and the way it was is more explicit IMO.

**Review by maxhutch — COMMENTED** [review]
That's a lot of new rules.  Can you confirm that none of the out key names have changed? I didn't see any, but there were a lot of changes.

**Comment by hegdevinayi** [comment]
> That's a lot of new rules. Can you confirm that none of the out key names have changed? I didn't see any, but there were a lot of changes.

Good point. I did not change any out key names. In some cases, there are additional keys referring to the same quantity, e.g. the "old" key for the energy convergence criterion/ionic step is `energy criteria`, which still exists, but there is a new key `ionic energy convergence threshold` that has the same value.

File: dftparse/pwscf/stdout_parser.py (hunk 1 above)
Comment (hegdevinayi) [reply dftparse/pwscf/stdout_parser.py]: This may not be an issue for this package as of now (no multiple inheritance), but following the discussion here https://stackoverflow.com/questions/576169/understanding-python-super-with-init-methods (the second answer), i think using `super().__init__()` is just better?

Comment (maxhutch) [reply dftparse/pwscf/stdout_parser.py]: If we have multiple inheritance, then we can explicitly call both init methods.  For example: https://github.com/CitrineInformatics/taurus/blob/develop/taurus/entity/object/measurement_spec.py#L16 .

From your link:
> But the main advantage comes with multiple inheritance, where all sorts of fun stuff can happen. 

`all sorts of fun` is the kind of phrase that makes me not want to use a feature unless we're really getting something valuable out of it.  Here, they are totally equivalent.

Comment (hegdevinayi) [reply dftparse/pwscf/stdout_parser.py]: Haha, fair. I'll change it back.

Comment (hegdevinayi) [reply dftparse/pwscf/stdout_parser.py]: My understanding is that using `super` instead of hardcoding the parent class names gives greater flexibility when subclassing (see the "dependency injection" section): https://stackoverflow.com/questions/222877/what-does-super-do-in-python/33469090#33469090. What do you think?

Comment (maxhutch) [reply dftparse/pwscf/stdout_parser.py]: If we decide to do DI, then we can change it.  Until that point, it is layer of indirection that serves no value.

**Review by maxhutch — APPROVED** [review]
